src/hoa-rotator.js

```diff
@@ -0,0 +1,230 @@
+/**
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+
+/**
+ * @fileOverview Sound field rotator for higher-order-ambisonics decoding.
+ */
+
+
+/**
+ * @class Higher-order-ambisonic decoder based on gain node network.
+ *        We expect the order of the channels to conform to ACN ordering.
+ * @param {AudioContext} context    Associated AudioContext.
+ * @param {Number} order            Ambisonic order (2 or 3).
+ */
+function HOARotator (context, order) {
+  this._context = context;
+
+  var num_channels = (order + 1) * (order + 1);
+  this._splitter = this._context.createChannelSplitter(num_channels);
+  this._merger = this._context.createChannelMerger(num_channels);
+
+  // Create a set of per-order rotation matrices using gain nodes.
+  this._m = Array(order);
+  for (var l = 1; l <= order; l++) {
+    var order_offset = l * l;
+    var rows = (2 * l + 1);
+    this._m[l - 1] = Array(rows * rows);
+
+    for (var j = 0; j < rows; j++) {
+      var input_index = order_offset + j;
+      for (var k = 0; k < rows; k++) {
+        var output_index = order_offset + k;
+        var matrix_index = j * rows + k; // Row-wise indexing.
+
+        this._m[l - 1][matrix_index] = this._context.createGain();
+        this._splitter.connect(this._m[l - 1][matrix_index], input_index);
+        this._m[l - 1][matrix_index].connect(this._merger, 0, output_index);
+      }
+    }
+  }
+
+  // W-channel requires not transform.
+  this._splitter.connect(this._merger, 0, 0);
+
+  // Default Identity matrix.
+  this.setRotationMatrix([1, 0, 0, 0, 1, 0, 0, 0, 1]);
+
+  // Input/Output proxy.
+  this.input = this._splitter;
+  this.output = this._merger;
+};
+
+HOARotator.prototype._kroneckerDelta = function(i, j) {
```

Comment:
Also this can be a local static function rather than a part of this class definition. Less exposure is better.